### DISCUSSION TITLE
New run form improvements

### DISCUSSION THREAD

**Discussion by jmdetloff** [biosimulations/biosimulations#4296] [Ideas]
I noticed a few things that could be improved about the new run UI that weren't within the scope of #4295 I'm not sure if all are high pri enough to merit an issue at the moment, but thought I'd point them out for the future.

- The url field currently has no debouncing. Every character that is typed into the url field initiates a new request to the field's value to attempt to download an archive. Most users probably paste in a url and in that case this is not an issue, but if a user typed in a long url the app would attempt and fail dozens of requests before the full url is typed in. To repro this open the form and tap a few characters into the url field and watch the 400's appear in the console.
- The networking code in dispatch.component.ts does not account for the possibility of multiple network calls. The component posts an archive and downloads the sed spec as soon as a project is entered, without waiting for form submission, so if a user enters multiple urls one after another the UI could enter a broken state where multiple subscription handlers are being called, potentially out of order.
- There isn't currently any UI indication if posting the selected archive or downloading the sed spec for the posted archive fails.
- The form should (maybe) use either linkified text or link icons, as opposed to both. Is there an intention to use linkified text vs the icons in a particular scenario?

**Comment by jonrkarr** [comment]
> The url field currently has no debouncing.

Correct. We overlooked this. This is easy to fix.

Issue: #4298

> The networking code in dispatch.component.ts does not account for the possibility of multiple network calls.

Yes, we should cancel the earlier calls.

Issue: #4299 

> There isn't currently any UI indication if posting the selected archive or downloading the sed spec for the posted archive fails.

I'm not sure what error you're referring to. We're using snack bars to display errors with posting model files. See https://github.com/biosimulations/biosimulations/blob/6a907085e2c9e8f2ba9a3b10bee7709909cc0aa5/libs/simulation-project-utils/ui/src/lib/create-project/create-project/create-project.component.ts#L715

Issue: #4300 

> The form should (maybe) use either linkified text or link icons, as opposed to both. Is there an intention to use linkified text vs the icons in a particular scenario?

We should be more consistent about this. The GUI often uses icon, but isn't consistent. The GUI could go either way to be internally consistent. (For completeness, the documentation uses linked text.)

Issue: #4301

**Reply by bilalshaikh42** [reply]
> 'm not sure what error you're referring to. We're using snack bars to display errors with posting model files. See

This refers to errors that may happen when the combine archive is first uploaded to the combine-api to do model introspection or when the network call to get available simulators/algorithm fails. We do have an error when the submission fails, but I am not sure about the intermediate steps

**Reply by jonrkarr** [reply]
The snack bar is also used to display a message on failures to get simulators/algorithms.

**Reply by jonrkarr** [reply]
The first two are easy to fix. I'll hold off to avoid creating merge conflicts with #4295.

**Comment by jmdetloff** [comment]
The potential error I'm referring to is when the call into combineApiService.getSpecsOfSedDocsInCombineArchive might fail. The dispatch component doesn't register any error handling here, and the combineApiService is catching errors but doesn't look like it's presenting anything user facing (that I can tell). In this potential error case users wouldn't see their fields populate with data from the sed spec, not a terrible experience, so I'm not sure if a user visible error is appropriate here but thought I'd call it out in case.

**Reply by jonrkarr** [reply]
Thanks for the careful attention to detail! I'm impressed. The UI is probably the least organized part of the code, (other than the code for wrangling heterogeneous models and simulation tools).
